Logging: Overhaul

nordic-krch opened the pull request:
Logging v2 highlights
----------------------------
- `log_strdup` no longer needed!!
- logging `long long`, `float` and `double` is now possible!!
- hex dumps with formatted strings
- Self-contained messages (even transient strings are there)
- 32/64 bit timestamp
- better memory utilization - ~20% more messages fit in the logging buffer
- 6x faster logging from userspace!
- overall simplified implementation

API changes
-----------------
No change in logging API! Add `CONFIG_LOG2_MODE_DEFFERED=y` to run sample with new version. Only backend API extended with a function for processing v2 messages. PR updates following backends to support v2:
- uart
- shell
- adsp
- rtt
- native_posix

To be added later: bluetooth monitor, net, spinel, swo, xtensa sim

Internals
------------
It is utilizing cbprintf packages (#30675, #31102) to create self-contained log message. Using `_Generic` feature of C11, it is possible to act based on types of arguments. It is used to determine when formatted string requires time consuming, runtime package creation and when log message can be created as simple serialization of the arguments. When  `char *` arguments are present in the log message then package is created at runtime. This includes copying into a message any string argument which is from rw memory.

Additionally, log messages are no longer stored as fragmented chunks allocated from a memslab. They are stored in a circular buffer. This ensures that log message is not fragmented and improves memory utilization. Circular buffer is somehow generic multi producer, single consumer packet buffer (`lib/os/mpsc_pbuf`). Additional requirement is to support overwrite mode where oldest packets are dropped when new message cannot be allocated. MPSC packet buffer uses alloc, commit, claim, free sequence to avoid copying. Additionally, it has functions optimized for storing single word and single word + pointer packets. It is forseen that those will be used by the tracing to logger functionality.

It pends currently on #33567 being solved. PR contains temporary workaround to get CI green.

PR content
---------------
- mpsc packet buffer
- new log_msg2 implementation
- adaptation of the logger to support both version
- adaptation of  some logger backends to support v2 api

Fixes #26170
Fixes #18351
Fixes #26246
Fixes #30353

nordic-krch committed:
logging: Add logging v2 implementation for log_msg

Added implementation of log_msg2 which is creating log messages
using cbprintf packaging and storing them in circular ring buffer
(mpsg_pbuf).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: log_output: Extend to support log_msg2 parsing

Added parsing of log_msg2.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: logging: Add test for log_msg2

Added suite for log_msg2 macros.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Adapt logger to support both versions

Adapted to be able to switch between v1 and v2.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: backend_uart: Adapt to support v2 api

Adapted to handle processing of v2 log messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
lib: os: Add mpsc (multiple producer, single consumer) packet buffer

Added module for storing variable length packets in a ring buffer.
Implementation assumes multiple producing contexts and single consumer.

API provides zero copy functionality with alloc, commit, claim, free
scheme.

Additionally, there are functions optimized for storing single word
packets and packets consisting of a word and a pointer. Buffer can work
in two modes: saturation or overwriting the oldest packets when buffer
has no space to allocate for a new buffer.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: Add test for mpsc_pbuf

Added suite for multiple producer, single consumer packet buffer.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add v2 support to native_posix backend

Add support for logging v2 to native_posix backend.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add v2 support to RTT backend

Add support for logging v2 to RTT backend.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Refactor in preparation for logging v2

Refactor and cleanup in preparation for v2.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Adapt ADSP backend to v2

Added support for v2 backend interface to ADSP.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Add support for logging v2

Extended shell to support logging v2.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add LOG_PRINTK macro for printing via logging

Added macro which allows to print formatted string using
logging infrastructure.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

andyross reviewed (commented):
Haven't done much more than a cursory scan, but this looks pretty clean to my eyes.

One question I can't find an immediate answer to: is this still running the backends out of a timer vs. letting them pull the output as they have bandwidth?  It looks like it might be, because I don't see anything like that removed.  This has been one of my pet peeves: backends with fast transport[1] drop a LOT of messages essentially needlessly, simply because the logging core never even tries to write them.

[1] The one I work on the most right now is the intel_adsp, where trace output goes into a region of system DRAM shared with the host OS -- it can sink bytes as fast as the DSP can write them)

nordic-krch commented:
@andyross following approach is used: logging operates with a circular packet buffer (contrary to list of fragmented messages allocated from memslab). When logging is processing it passes the pointer to the message from the packet buffer to each backend. Call to the backend is synchronous so logging expects that when it returns message is no longer used by the backend. Backend has two option: block the logging thread and process the message or copy the message to its own memory and return.

nordic-krch committed:
lib: os: mpsc_pbuf: Add optional debug features

Added optional debug prints. Logging cannot be used because
mpsc pbuf is used by the logging.

Added option to clear packet memory after allocation. Option is
enabled in Kconfig.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
bluetooth: Add cast to forward declared struct

Logging v2 is using _Generic keyword for detecting type of
log message arguments. Apparently, it does not support handling
of pointers to forward declared structures. Added casting to void *.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
net: l2: bluetooth: Add cast to forward declared struct

Logging v2 is using _Generic keyword for detecting type of
log message arguments. Apparently, it does not support handling
of pointers to forward declared structures. Added casting to void *.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
manifest: Temporary point to sof repo with MAX workaround

Point to commit which contains temporary fix for sof redefining
MAX macro.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: logging: Add benchmark test

Added test which benchmarks deferred logging v1 and v2.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: misc: formatted_output: Add section for cbprintf packaging

Add documentation for cbprintf packaging.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: misc: Add documentation for MPSC Packet Buffer

Add documentation for Multi Producer Single Consumer Packet Buffer.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: logging: Add documentation for logging v2

Extended and refactored logging documentation. Added details about
logging v2 and comparison with v1.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
PR is ready for review. I recommend to read logging documentation first as it highlights the differences between v1 and v2.

jukkar commented on include/sys/mpsc_pbuf.h:
```diff
@@ -0,0 +1,237 @@
+
+/** @brief MPSC packet buffer structure. */
+struct mpsc_pbuf_buffer {
+	/* Temporary write index. */
```
The individual fields should have documentation comment added too (if you want these to be present in the docs) so `/** xxxx` should written here. If you want to hide some fields from documentation, mark them separately (I am not sure what is the correct way, but in networking I have used `/** @cond INTERNAL_HIDDEN */` and `/** @endcond */` pattern).

jukkar commented on include/sys/mpsc_pbuf.h:
```diff
@@ -0,0 +1,237 @@
+	/* Read index. */
+	uint32_t rd_idx;
+
+	uint32_t flags;
```
All fields should have a document comment, or you need to hide them separately. The generated documentation will look a bit unfriendly to the user reading the docs now.

jukkar commented on lib/os/mpsc_pbuf.c:
```diff
@@ -0,0 +1,399 @@
+	union mpsc_pbuf_generic *dropped_item = NULL;
+
+	do {
+
```
extra empty line can be removed

jukkar commented on lib/os/mpsc_pbuf.c:
```diff
@@ -0,0 +1,399 @@
+	uint32_t free_wlen;
+
+	do {
+		cont = false;
```
better would be to put the cont after the variables are introduced (so three lines after this one).

jukkar commented on lib/os/mpsc_pbuf.c:
```diff
@@ -0,0 +1,399 @@
+
+	do {
+
+		cont = false;
```
same here, and also extra empty line is unncessary

jukkar commented on lib/os/mpsc_pbuf.c:
```diff
@@ -0,0 +1,399 @@
+	union mpsc_pbuf_generic *dropped_item = NULL;
+
+	do {
+		cont = false;
```
ditto

jukkar commented on lib/os/mpsc_pbuf.c:
```diff
@@ -0,0 +1,399 @@
+	bool wrap;
+
+	do {
+		cont = false;
```
ditto

nordic-krch commented:
@jukkar I've applied your comments.

@dcpleung I've updated documentation with description of log message and cbprintf package content.

nordic-krch commented:
@nashif @dcpleung I've added dedicated section for log strings `log_strings_sections`. I've been playing a bit with `objcopy` and tried to remove that section but removing is just filling given section with 0's. It would need to be placed at the end probably. Or maybe you have some other ideas.

dcpleung commented:
@nordic-krch Thank you. I have updated #30539 on top of this.

As for removing the section, it will probably require some linker magic. For example, on boards that is flashed with a .bin file, the log string section can be put into non-existing memory (thus won't appear in the final .bin file). But it still exists in the ELF file so we can extract the strings. On QEMU, this won't work because it will try to copy everything from ELF file into memory.

carlescufi commented:
> @nordic-krch Thank you. I have updated #30539 on top of this.

@dcpleung could you also add a review to this PR so we can make sure it gets enough reviewers in time for 2.6?
@andyross, @jukkar, @joerchan, @jhedberg @Vudentz it would be great to get more reviews.

dcpleung commented on lib/os/mpsc_pbuf.c:
```diff
@@ -5,6 +5,26 @@
 
+#define MPSC_PBUF_DEBUG 0
+
+#define MPSC_PBUF_DBG(buffer, ...) do { \
+	if (MPSC_PBUF_DEBUG) { \
+		printk(__VA_ARGS__); \
+		if (buffer) { \
+			mpsc_state_print(buffer); \
+		} \
+	} \
+} while (0)
+
+static inline void mpsc_state_print(struct mpsc_pbuf_buffer *buffer)
+{
+	if (MPSC_PBUF_DEBUG) {
+		printk("wr:%d/%d, rd:%d/%d\n",
+			buffer->wr_idx, buffer->tmp_wr_idx,
+			buffer->rd_idx, buffer->tmp_rd_idx);
+	}
+}
```
Do these require `CONFIG_LOG_PRINTK` to be disabled?

nordic-krch replied:
no, when `CONFIG_LOG_PRINTK` is enabled then this module is not used by the logger so it can printk without issue. I've added it for debugging purposes during testing. Since default logging relies on it, i didn't want to use the logger.

nordic-krch committed:
net: lib: openthread: logging: Refactor logging function

Refactor logging function to avoid multiple macro calls.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

dcpleung commented:
When long double is used but `CONFIG_CBPRINTF_PACKAGE_LONGDOUBLE=n`, there are bunch of compilation errors on the log macros themselves, and the error message about needing to enable the kconfig is buried among others. Just wondering if it is possible to get rid of the other errors with some macro magic?

Also, long double does not seem to work. `%Lf` is printed literally, but not the value. Maybe some other kconfig is needed besides `CONFIG_CBPRINTF_PACKAGE_LONGDOUBLE`?

pabigot commented:
> Also, long double does not seem to work. `%Lf` is printed literally, but not the value. Maybe some other kconfig is needed besides `CONFIG_CBPRINTF_PACKAGE_LONGDOUBLE`?

cbprintf_complete doesn't format long double, because there's no standard definition of how many bits it is and so there's no way to interpret it.  It gets flagged as unsupported when the length modifier is parsed.

nordic-krch commented:
> Just wondering if it is possible to get rid of the other errors with some macro magic?

Unfortunately, i don't see a way to do that as each argument in any log must be analyzed there is no way to avoid for loop macro and that results in the nested compiler errors. I would say that we are lucky that using `_Generic` we can detect that at compile time.

carlescufi commented:
@dcpleung as part of the work you are doing with dictionary-based logging, which needs to be coupled to this overhaul, could you please also review this PR and approve if appropriate?

dcpleung commented on include/linker/common-rom.ld:
```diff
@@ -153,6 +153,13 @@
+		__log_strings_start = .;
+		KEEP(*(SORT(.log_strings*)));
+		__log_strings_end = .;
+	} GROUP_LINK_IN(ROMABLE_REGION)
```
This needs to be `GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)` due to linker magic needed for `CONFIG_DEMAND_PAGING'. This will fix the issue with build the board `qemu_x86_virt`.

nordic-krch replied:
thanks. Fixed.

dcpleung reviewed (approved):
Just need to fix the build error and it should be good to go.

dcpleung commented:
I am getting errors while running twister on `qemu_leon3`. They all failed like this:
```
START - test_mode_size_str_with_2strings

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/logging/log_msg2/src/main.c:344: get_msg_validate_length: (len not equal to exp_len)
Unexpected message length 8 (exp:14)
```
for tests under `tests/subsys/logging/log_msg2`. CI seems to have no issue so just wondering what might be an issue.

SwissKnife64 commented:
Here is the list of the required CONFIG to get the LOG2 for floating point working.

**It is mandatory to set `CONFIG_CBPRINTF_FP_A_SUPPORT=y`** to get `%f` and `%g` working! And `%a` is working also ;-).

```
CONFIG_LOG=y
CONFIG_LOG2_MODE_DEFERRED=y

CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y

CONFIG_CBPRINTF_FP_A_SUPPORT=y

```

npitre commented:
> **It is mandatory to set `CONFIG_CBPRINTF_FP_A_SUPPORT=y`** to get `%f` and `%g` working! And `%a` is working also ;-).

No. CONFIG_CBPRINTF_FP_A_SUPPORT is for ... %a.

%f and %g just need CONFIG_CBPRINTF_FP_SUPPORT.

nordic-krch commented:
Just a note, `CONFIG_CBPRINTF_FP_SUPPORT` defaults to `y` when `CONFIG_FPU` is enabled so when FPU is in use you shouldn't need any additional explicit configuration.

SwissKnife64 commented:
The warnings are gone, may be they were present with an outdated master on my machine.

Here is my CMakeList.txt

```
# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.13.1)

# START added for CLion
set(ENV{BOARD} nucleo_h743zi)
set(ENV{ZEPHYR_SDK_INSTALL_DIR} /usr/local/zephyr-sdk-0.12.3)
set(ENV{ZEPHYR_TOOLCHAIN_VARIANT} gnuarmemb)
[Output.txt](https://github.com/zephyrproject-rtos/zephyr/files/6456915/Output.txt)

#toolchain is: /usr/local/gcc-arm-none-eabi-10-2020-q4-major/arm-none-eabi
# STOP added for CLion

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(blinky)

target_sources(app PRIVATE src/main.c)

```
